quinn-udp/src/windows.rs

```diff
@@ -68,38 +133,162 @@ impl UdpSocketState {
             if rc == -1 {
                 return Err(io::Error::last_os_error());
             }
+
+            if wsa_recvmsg_ptr.is_some() {
+                let rc = unsafe {
+                    WinSock::setsockopt(
+                        socket.0.as_raw_socket() as _,
+                        WinSock::IPPROTO_IPV6 as _,
+                        WinSock::IPV6_PKTINFO as _,
+                        &sock_true as *const _ as _,
+                        mem::size_of_val(&sock_true) as _,
+                    )
+                };
+                if rc == -1 {
+                    return Err(io::Error::last_os_error());
+                }
+
+                let rc = unsafe {
+                    WinSock::setsockopt(
+                        socket.0.as_raw_socket() as _,
+                        WinSock::IPPROTO_IPV6 as _,
+                        WinSock::IPV6_ECN as _,
+                        &sock_true as *const _ as _,
+                        mem::size_of_val(&sock_true) as _,
+                    )
+                };
+                if rc == -1 {
+                    return Err(io::Error::last_os_error());
+                }
+            }
         }
 
         let now = Instant::now();
         Ok(Self {
             last_send_error: Mutex::new(now.checked_sub(2 * IO_ERROR_LOG_INTERVAL).unwrap_or(now)),
+            wsa_recvmsg_ptr,
         })
     }
 
     pub fn send(&self, socket: UdpSockRef<'_>, transmits: &[Transmit]) -> Result<usize, io::Error> {
         let mut sent = 0;
         for transmit in transmits {
-            match socket.0.send_to(
-                &transmit.contents,
-                &socket2::SockAddr::from(transmit.destination),
-            ) {
-                Ok(_) => {
-                    sent += 1;
+            // FIXME be symmetric ? do not send ECN if we cannot receive them ?
+            if self.wsa_recvmsg_ptr.is_some() {
```

Comment:
Maybe move the body of this into a helper function for clarity and to limit indentation?

Reply:
ok i will do it